### ISSUE TITLE
Laguerre second parameter f64

### ISSUE THREAD

**Issue by nilgoyette** [At0micBee/scilib#10]
Bonjour. I'm trying to port a python script to Rust and I have a problem on this call

    from scipy.special import genlaguerre
    ...
    genlaguerre(usize, f64)(list of f64)

AFAIK, you have the only Laguerre Rust implementation so I added your library to my project. I realized that your implementation works for

    let v = Laguerre::new(usize, i32).compute(f64);

You both have the same answers when I use an integer for the second parameter. However, your implementation is **much** cleaner! Bravo :)

I'm not a mathematician myself; I simply use their tools, so I don't really know the complexity of what I am asking or if it makes sense. Would it be complex to support a floating point number for the second parameter? What [scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.genlaguerre.html) calls "alpha"?

**Comment by At0micBee** [comment]
Hi there!

Thanks for using this library and for the kind words, very much appreciated!

I have had a quick look into the way `scipy` does things, and I should be able to implement it. This will required adding a few other functions, as some experimenting will be required on my part to make sure everything works, and remains fast. I originally needed those function for an old IDL script that I was rewriting, and IDL doesn't use the `float` option, but rather an integer, [see here](https://www.l3harrisgeospatial.com/docs/laguerre.html).

It is however a good idea to add this as an option, thank you again for letting me know. I'll try to get on it during the week and push a patch ASAP. If I haven't given an update by next week-end feel free to contact me!

**Comment by At0micBee** [comment]
Okay, update time!

I've added a generalized form of the binomial coefficients, which allows for a real input. Since this is the generating method of the Laguerre coefficients, we can use any real value for the alpha parameter. In the example I've added in the docs, I've added a test with alpha = -1.5, and the coefficients (and result) are confirmed by wolfram alpha.

Since I was on it, I've remodeled the way polynomial are created, and I've constructed a single struct, `Poly`, which serves as a building block for any of them. All the previous polynomial now fall in this format. I've added a couple new poly while I was at it, and I'm going to finish this new version properly by the end of the week-end.

Bottom line is that the real value of alpha is possible, and since it's part of the generating function, the struct can be entirely generated to compute multiple values, without needing expensive computation every time. E.g., as before, the following works:
```rust
let la: Poly = Poly::laguerre(5, -0.74);
let v1 = la.compute(7.2);
let v2 = la.compute(-1.2);
let v3 = la.compute_complex(Complex64::new(1.2, -0.02);
// etc...
```

I'll try to wrap up the remaining additions I've done today and put the patch online early next week, as version `0.6.0`.

**Comment by nilgoyette** [comment]
Wow, thank you for your fast response work!

If I may ask a question. Why do you have 2 methods: `compute` and `compute_complex`? The code for both in "identical"

    self.factors.iter().zip(&self.powers).fold(                 0.0, |res, (f, p)| res +  f * x.powi(*p))
    self.factors.iter().zip(&self.powers).fold(Complex64::default(), |res, (f, p)| res + *f * z.powi(*p))

Can't you use a single function with a trait, like `T: Default`?

**Comment by At0micBee** [comment]
The short answer is that the `.powi(n)` function doesn't have a shared trait between them. This means that just adding the `Default` trait will not allow the function to work.

It should be possible to merge them by using traits from the `num_traits` crate, but so far I haven't had the time to properly check everything.

**Comment by nilgoyette** [comment]
> The short answer is that the `.powi(n)` function doesn't have a shared trait between them.

Oh, ok, my bad. I only saw the "default" part.

**Comment by At0micBee** [comment]
Alright, I just pushed version `0.6.0` with the new and improved polynomials. The tests I ran on my side were all passing so I'll be closing this issue. If you experience any problems with it, feel free to open a new one.

Cheers,
AB.
